products/facilities/regional-offices/engineering/testing-qa.md

```diff
@@ -1 +1,90 @@
+
+
+## Test Plan
+1. Since the contents of the VBA Regional Office Pages are dynamic (they can be created, removed, updated of all content, etc. by CMS editors), the contents of the pages cannot be tested by e2e tests.
+2. Since the majority of the contents of the VBA Regional Office Pages are generated at build time using content-build, cypress e2e tests with mocking data cannot be used to test.
+3. We use a combination of manual content-releases of VBA content on tugboats and automated Continuous Integration-based unit and component testing to verify that the segments of liquid templates render into correct HTML pages (does not allow for rendering React components). For React components we use regular component testing in the vets-website repository.
+   
+## Manual Testing
+## Automated testing
+1. The basis of automated tests for VBA pages use mock data (or snapshot-in-time data from CMS) that covers all template pieces that can be visible in a VBA page.
+2. We run tests using the mocha test runner system.
+3. A mocha test uses the `renderHTML` method that replicates the metalsmith process of building pages from drupal content. This allows us to build tests that verify the HTML structure of the page that would be the same as one built from downloaded CMS data.
+    1. For example: since we wanted to make sure multiple banner types can appear on a page, we build a CMS node of a VBA facility with a banner alert and take a segment of the generated JSON from Drupal. We repeat this for all banner types. Then we use renderHTML to build the template pieces and 'expect' that each banner type appears when the mock data is present for it. Since banner templates are built on `<va-banner>` components, we are only able to test props of the Design System Components.
+4. There is a limitation to testing via renderHTML because it does not pull in or load react components and it does not allow us to test deep in the shadow DOM. So for all Design System Components (which we rely on) we expect that the Design System Team has tested for certain accessibility  features. Our manual checks also verify the Design System Components are acting as expected.
+5. To facilitate processing the drupal content into templated pages, we often use liquid filters. Whenever we create a liquid filter we add tests to the CI pipeline for that filter. We always cover null conditions and valid object types (because all data from Drupal comes as objects). You will see below the list of liquid filters added and the tests created.
+
+## Regression Testing
+Regression testing is built into the automated testing and the CI will not succeed in building and deploying the content if it does not comply with the automated testing. However, if CMS content is changed and is not congruent with the structure of the GraphQL the CI will not succeed. Testing should ensure for all potential options that an editor is presented with for the particular component in the VBA Regional Benefit Office pages.
```

Comment:
I'm having a hard time grokking this piece: 
> Testing should ensure for all potential options that an editor is presented with for the particular component in the VBA Regional Benefit Office pages.

Does this say the same thing you mean? 
Testing should ensure that all potential options an editor is presented with for the particular component in the VBA Regional Benefit Office pages are unchanged.

Reply:
Hmm, I think I meant that CMS structure (if changed) would produce an error if it changes something expected by the graphql.

Reply:
Oh, I re-read which question you were asking. I mean that if the editor is given the option to add a wysiwyg paragraph, a FAQ section, etc, that we should be able to test all combinations.

Reply:
I think the original line is still a bit confusing, but I hear where you're headed. What if that line is updated to: 

This approach to testing should ensure coverage for all potential data options or combinations that an editor is presented with inside the Drupal CMS, and should protect against unexpected data structure changes coming from Drupal, for the components in the VBA Regional Benefit Office pages.

Does that seem accurate?